### PR TITLE
Use gaussian elimination for positive-definite check

### LINKED ISSUE #18955: is_positive_definite gives wrong result
The is_positive_definite does not always give a result. In the example below with a 4x4 matrix, which numpy can demonstrate is posdef, sympy returns neither False nor True but None. None is interpreted as False in an if statement and the result thus becomes wrong.

```python
import numpy as np
import sympy

M = [[0.0228202735623867, 0.00518748979085398, -0.0743036351048907, -0.00709135324903921],
    [0.00518748979085398, 0.0349045359786350, 0.0830317991056637, 0.00233147902806909],
    [-0.0743036351048907, 0.0830317991056637, 1.15859676366277, 0.340359081555988],
    [-0.00709135324903921, 0.00233147902806909, 0.340359081555988, 0.928147644848199]]

M_sympy = sympy.Matrix(M)
print(M_sympy.is_positive_definite)

M_numpy = np.array(M)
print(np.linalg.cholesky(M_numpy))
```

### PR DESCRIPTION
**Pull request by sylee957**
#### References to other Issues or PRs

Fixes #18955

#### Brief description of what is fixed or changed

I have changed positive definite check to use division free gaussian elimination, which hopefully gives less bloated expressions, and used berkowitz determinant check for positive semidefinite check which can suffer less from zero testing.

#### Other comments

Screen Shot 2020-04-28 at 01 02 29
Screen Shot 2020-04-28 at 01 02 47

#### Release Notes

- matrices
  - Added `is_strongly_diagonally_dominant` and `is_weakly_diagonally_dominant` properties for `Matrix`.

### CODE REVIEW

File: sympy/matrices/tests/test_matrices.py [hunk 1]
```diff
@@ -1029,56 +1029,6 @@ def test_definite():
     assert m.is_indefinite == True
 
 
-def test_positive_definite():
-    # Test alternative algorithms for testing positive definitiveness.
-    m = Matrix([[2, -1, 0], [-1, 2, -1], [0, -1, 2]])
-    assert m._eval_is_positive_definite(method='eigen') == True
-    assert m._eval_is_positive_definite(method='LDL') == True
-    assert m._eval_is_positive_definite(method='CH') == True
```
Comment (oscarbenjamin) [inline sympy/matrices/tests/test_matrices.py:1037]: Are these tests not needed any more?

Comment (sylee957) [reply sympy/matrices/tests/test_matrices.py:1037]: I have removed those methods so they cannot be tested

**Comment by oscarbenjamin** [comment]
If the matrix is not square should we say that `is_positive_definite -> False`? What about if it is not Hermitian?

**Comment by sylee957** [comment]
I have left them undefined because I found nobody had asserted that any nonsquare or nonhermitian matrix is not positive definite. 

But I have found a generalization that can be extended for nonhermitian matrices in https://www.jstor.org/stable/2317709, which also notes about some properties like the existance of the LR factorization or the positive eigenvalues.

Though `False` can be asserted for nonsquare matrices, however, I don't think that it loses such triviality because checking whether a matrix is hermitian or square can be done parallel with `is_hermitian` or `is_square`.

**Comment by oscarbenjamin** [comment]
That paper has other useful properties for example that if a matrix positive definite then the diagonal entries are positive (not sure if this only applies for real matrices).

The contrapositive leads to a much cheaper check of positive definiteness: if any diagonal entry is not positive then the matrix is not positive definite.

**Comment by oscarbenjamin** [comment]
It's probably worth having a shortcut for the case where the answer is probably unknowable. The original issue here came from a symmetric matrix where every entry was a symbol with no assumptions.

I think that if we literally don't know whether any element on the diagonal is positive or not then it's very unlikely that we will be able to get determine if the matrix as a whole is positive definite e.g.:
```
if all(self[i, i].is_positive is None for i in range(N)):
    return None
```
Can you think of any examples where we don't know if any of the diagonals is positive or not but we do know that the matrix is or is not positive definite?

**Comment by oscarbenjamin** [comment]
Actually I guess this would be an example:
```julia
In [9]: x = Symbol('x', real=True)                                                                                                             

In [10]: M = Matrix([[sin(x), 2], [2, sin(x)]])                                                                                                

In [11]: M                                                                                                                                     
Out[11]: 
⎡sin(x)    2   ⎤
⎢              ⎥
⎣  2     sin(x)⎦
```
And in this example we don't know whether anything is positive or negative but we know at least one eigenvalue is negative either way:
```julia
In [21]: M = Matrix([[sin(x), 2*sin(x)], [2*sin(x), sin(x)]])                                                                                  

In [22]: M                                                                                                                                     
Out[22]: 
⎡ sin(x)   2⋅sin(x)⎤
⎢                  ⎥
⎣2⋅sin(x)   sin(x) ⎦

In [23]: M.eigenvals()                                                                                                                         
Out[23]: {-sin(x): 1, 3⋅sin(x): 1}
```

**Comment by oscarbenjamin** [comment]
Looks good. Merging

**Comment by oscarbenjamin** [comment]
What exactly is the relationship between row and column diagonal dominance? Are they equivalent?

**Comment by sylee957** [comment]
They are trivially equivalent for symmetric or hermitian matrices which is sufficient for this test. For other cases I'm not familiar enough to assert.

**Pull request merged**